File: config-service/src/config_service/models/application.py
"""Pydantic models for Application entity."""

from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field, ConfigDict, field_validator
import re


class ApplicationBase(BaseModel):
    """Base application model with common fields."""
    name: str = Field(..., min_length=1, max_length=256, description="Application name")
    comments: Optional[str] = Field(None, max_length=1024, description="Application comments")


class ApplicationCreate(ApplicationBase):
    """Model for creating a new application."""
    pass


class ApplicationUpdate(ApplicationBase):
    """Model for updating an existing application."""
    pass


class Application(ApplicationBase):
    """Complete application model with all fields."""
    model_config = ConfigDict(from_attributes=True)
    
    id: str = Field(..., description="Application unique identifier (ULID format)")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")
    
    @field_validator('id')
    @classmethod
    def validate_ulid(cls, v: str) -> str:
        """Validate that the ID is a valid ULID format."""
        if not re.match(r'^[0-9A-HJKMNP-TV-Z]{26}$', v):
            raise ValueError('Invalid ULID format')
        return v


class ApplicationWithConfigs(Application):
    """Application model including related configuration IDs."""
    configuration_ids: List[str] = Field(default_factory=list, description="Related configuration IDs (ULID format)")
    
    @field_validator('configuration_ids')
    @classmethod
    def validate_config_ulids(cls, v: List[str]) -> List[str]:
        """Validate that all configuration IDs are valid ULID format."""
        for ulid in v:
            if not re.match(r'^[0-9A-HJKMNP-TV-Z]{26}$', ulid):
                raise ValueError(f'Invalid ULID format: {ulid}')
        return v

File: config-service-dotnet/src/ConfigService.Core/Models/Application.cs
using System.ComponentModel.DataAnnotations;
using System.Text.RegularExpressions;

namespace ConfigService.Core.Models;

/// <summary>
/// Base application model with common fields
/// </summary>
public abstract class ApplicationBase
{
    [Required]
    [StringLength(256, MinimumLength = 1)]
    public string Name { get; set; } = string.Empty;

    [StringLength(1024)]
    public string? Comments { get; set; }
}

/// <summary>
/// Model for creating a new application
/// </summary>
public class ApplicationCreate : ApplicationBase
{
}

/// <summary>
/// Model for updating an existing application
/// </summary>
public class ApplicationUpdate : ApplicationBase
{
}

/// <summary>
/// Complete application model with all fields
/// </summary>
public class Application : ApplicationBase
{
    private static readonly Regex UlidRegex = new(@"^[0-9A-HJKMNP-TV-Z]{26}$", RegexOptions.Compiled);

    [Required]
    public string Id { get; set; } = string.Empty;

    [Required]
    public DateTime CreatedAt { get; set; }

    [Required]
    public DateTime UpdatedAt { get; set; }

    /// <summary>
    /// Validates that the ID is a valid ULID format
    /// </summary>
    public bool IsValidUlid() => UlidRegex.IsMatch(Id);

    /// <summary>
    /// Validates a ULID string format
    /// </summary>
    public static bool IsValidUlid(string ulid) => !string.IsNullOrEmpty(ulid) && UlidRegex.IsMatch(ulid);
}

/// <summary>
/// Application model including related configuration IDs
/// </summary>
public class ApplicationWithConfigs : Application
{
    public List<string> ConfigurationIds { get; set; } = new();

    /// <summary>
    /// Validates that all configuration IDs are valid ULID format
    /// </summary>
    public bool AreConfigurationIdsValid() => ConfigurationIds.All(IsValidUlid);
}